spec/02-integration/05-proxy/01-router_spec.lua

```diff
@@ -387,4 +387,206 @@ describe("Router", function()
+        local i = 1
+
+        if strip_uri == true then
+          i = i + 3
+        elseif strip_uri == false then
+          i = i + 6
+        end
+
+        if upstream_uri ~= "/" then
+          i = i + 1
+          if string.sub(upstream_uri, -1) == "/" then
+            i = i + 1
+          end
+        end
+
+        local res = assert(client:send {
+          method  = "GET",
+          path    = request_uri,
+          headers = {
+            ["Host"] = "localbin-" .. i .. ".com",
+          }
+        })
+
+        local json = assert.res_status(200, res)
+        local data = cjson.decode(json)
+
+        assert.equal(expected_uri, data.vars.request_uri)
+      end
+    end
+
+    local checks = {
+      { "/",         "/",         "/",                 nil   },
+      { "/",         "/foo/bar",  "/",                 nil   },
+      { "/",         "/foo/bar/", "/",                 nil   },
+      { "/foo/bar",  "/",         "/foo/bar",          nil   },
+      { "/foo/bar/", "/",         "/foo/bar/",         nil   },
+      { "/foo/bar",  "/foo/bar",  "/foo/bar",          nil   },
+      { "/foo/bar/", "/foo/bar",  "/foo/bar",          nil   },
+      { "/foo/bar",  "/foo/bar/", "/foo/bar/",         nil   },
+      { "/foo/bar/", "/foo/bar/", "/foo/bar/",         nil   },
+      { "/",         "/",         "/",                 true  },
+      { "/",         "/foo/bar",  "/",                 true  },
+      { "/",         "/foo/bar/", "/",                 true  },
+      { "/foo/bar",  "/",         "/foo/bar",          true  },
+      { "/foo/bar/", "/",         "/foo/bar/",         true  },
+      { "/foo/bar",  "/foo/bar",  "/foo/bar",          true  },
+      { "/foo/bar/", "/foo/bar",  "/foo/bar",          true  },
+      { "/foo/bar",  "/foo/bar/", "/foo/bar/",         true  },
+      { "/foo/bar/", "/foo/bar/", "/foo/bar/",         true  },
+      { "/",         "/",         "/",                 false },
+      { "/",         "/foo/bar",  "/foo/bar",          false },
+      { "/",         "/foo/bar/", "/foo/bar/",         false },
+      { "/foo/bar",  "/",         "/foo/bar",          false },
+      { "/foo/bar/", "/",         "/foo/bar/",         false },
+      { "/foo/bar",  "/foo/bar",  "/foo/bar/foo/bar",  false },
+      { "/foo/bar/", "/foo/bar",  "/foo/bar/foo/bar",  false },
+      { "/foo/bar",  "/foo/bar/", "/foo/bar/foo/bar/", false },
+      { "/foo/bar/", "/foo/bar/", "/foo/bar/foo/bar/", false },
+    }
+
+    local unpack = table.unpack or unpack
+
+    for _, args in ipairs(checks) do
+
+      local config = "(strip_uri = n/a)"
+
+      if args[4] == true then
+        config = "(strip_uri = on) "
+      elseif args[4] == false then
+        config = "(strip_uri = off)"
+      end
+
+      local space = string.sub(args[1], -1) == "/" and "" or " "
+
+      it(config .. " is not appended to upstream uri " .. args[1] .. space .. " when requesting " .. args[2],
+        check(unpack(args)))
+    end
+  end)
```

Comment:
I feel like this test suite is *quite* complicated to maintain. Maybe more that just "quite", even.

Couldn't we come up with something more descriptive, like:

```lua
local checks = {
{
  api = {
    uris = { "/", "/foo/bar" },
    upstream_url = "http://localhost:9999/",
  },
  -- request_uri    expected_upstream_uri
  { "/",         "/" },
  { "/foo/",     "/foo/" },
},
{
  api = {
    uris = { "/foo/bar/" },
    upstream_url = "http://localhost:9999/foo",
  },
  -- request_uri    expected_upstream_uri
  { "/foo/",    "/foo/" },
  { "/foo",      "/foo" },
}
}
```

(with better alignment of the comments of course).

Also, let's note that an API with `uris=/,/foo/bar` does not really make sense in those tests since we are focusing on post-matching behavior. We should probably only declare APIs with `uris=/foo/bar` or `uris=/foo/bar/` with both `strip_uri=true|false`.

Doing so would have the benefit of: 
- grouping APIs declaration, input values and expected outputs together.
- adding a comment header to each value of the table that has declarative tests input/output values, so we know what they refer to.
- avoid the `i = i + n` logic in the `check()` closure, which is asks for a lot of cognitive load when debugging.
- we can still build a nice loop that generates tests and messages as already done
- it's easier to separate the use cases per API in the the `checks` table (currently they are all grouped together and we don't know which test belongs to which API without reading the `check()` closure.

Additionally, I strongly believe those tests should be part of the `01-unit` test suite, as per the router tests. They should most probably be moved in the `router_spec` tests over there, which already mocks the `ngx` object and allows for testing of the newly set URI (as currently already tested there). Since this feature extends the same part of the code, the tests should be over there. They will also be much faster, and avoid the need for extending the `custom_nginx.template` file (which was a quick fix but should be avoided as much as possible for now).